### ISSUE TITLE
If two shaders / permutations produce the same code, do we compile them twice?

### ISSUE THREAD

**Issue by illwieckz** [DaemonEngine/Daemon#1435]
If two shaders / permutations produce the same code, do we compile them twice?

**Comment by illwieckz** [comment]
If not, we may do an index of cached checksum, not just an index of shader+permutation id so we can deduplicate both compile time and storage.

**Comment by VReaperV** [comment]
The glsl caching code is a mess. It should be changed to be a straightforward "do all of the shader code processing, including adding macros, *then* check if we can load it based on that".

**Comment by VReaperV** [comment]
> If two shaders / permutations produce the same code, do we compile them twice?

Is there an example of such permutations or shaders? If two permutations exist for the same shader that are created by setting different macros, yet produce the same results, then some macros should just be nuked.

**Comment by illwieckz** [comment]
I wonder if that happens with generic2D.

**Comment by VReaperV** [comment]
> I wonder if that happens with generic2D.

You mean permutations of it?

**Comment by illwieckz** [comment]
Yes.

**Comment by illwieckz** [comment]
I mean I wonder if some permutations of generic2D and some permutations of generic may do the same.

**Comment by VReaperV** [comment]
> Yes.

There's only one: deph fade.

**Comment by VReaperV** [comment]
> I mean I wonder if some permutations of generic2D and some permutations of generic may do the same.

Well, technically generic2D shader always has:
```
#if defined(GENERIC_2D)
	gl_FragDepth = 0;
#endif
```

**Comment by VReaperV** [comment]
(the other one doesn't)

**Comment by VReaperV** [comment]
The above is no longer true, now that the generic2D shader no longer uses that and is basically the same as generic one now.

**Comment by VReaperV** [comment]
> If not, we may do an index of cached shader text checksum, not just an index of shader+permutation id so we can automatically deduplicate both compile time and storage.

This is probably the best option, especially considering some of the post-processing our shader sources go through, which is not cached properly I think.

**Comment by VReaperV** [comment]
Note that `generic2D` shader and the relevant macro no longer exist.

**Comment by VReaperV** [comment]
For only building the required shaders, see: https://github.com/DaemonEngine/Daemon/issues/1294#issuecomment-2378910816.

**Comment by slipher** [comment]
> If two shaders / permutations produce the same code, do we compile them twice?

Yes. Our shader cache files things under the shader name and permutation number, so it would never notice if a checksum is equal to a different shader's checksum.

Even if we did compare against all checksums, it is very unlikely that we would have two that are exactly identical because we operate on non-preprocessed code (well, except `#insert`). Even if we had a useless macro combination that would be identical to another one after preprocessing, the shader texts used for hashing would still be different because the macros themselves are part of the text that is (should be) hashed.

I will close this now. Feel free to create a new issue or reopen if there are any more questions.
